Make sure server selection stays in sync with server address, refactoring

Robyt3 opened the pull request:
- Make sure that the selected server is always in sync with the server address (handle Page change, Filter selection, Server selection, Filter change). Changing the address changes the selection, will now also deselect if no address matches. Closes #2003.
- Store selected filter for each browser page (internet, LAN) separately.
- Store selected server for each filter and browser page separately.
- Reenables the important message "No servers match your filter criteria" in case the filter excludes everything
- Adds additional important message "No filter category is selected" if no filter category is expanded, because this was really confusing the first time it happended.
- Adds additional important message "Fetching server info" when no servers are available because they are refreshing.
- Move CListBox class to menus_listbox.cpp
- Refactor CScrollRegion by moving the associated methods inside the class, move the class to menus_scrollregion.cpp
- Remove obsolete comments and some unused constants.
- Fixes browser headers selection not working because of clipped server entry behind (closes #2422). Also adds a background to the header to indicate hot state.
- Implement clicking on scrollbar of CScrollRegion to scroll there (closes #2429).

Dune-jr commented on src/game/client/components/menus.h:
```diff
@@ -628,21 +605,35 @@ class CMenus : public CComponent
-	};
+		const int Tab = ServerBrowser()->GetType();
+		if(m_aSelectedFilters[Tab] == -1)
+			return NULL;
```
I believe you are supposed to use `0` or `0x0` in Teeworlds instead of `NULL`. The rationale being that `NULL` is sometimes not zero.

Robyt3 replied:
Fixed it. That documentation file is also missing the prefix `_pfn` for function pointers.

Robyt3 commented:
I'll look into the configured selection being changed, there is probably a default somewhere where it shouldn't be.

I think we should split up some of the larger files (especially the menus and UI files) so it's more manageable and we can extract reuseable components (also to consolidate client and editor UI).

I'm not sure about the CRenderTools thou. It would improve the usage, but then CUIRect would have to know the graphics implementation. I think there is a broader discussion to be had about what should belong in ui, render and menus.

LordSk commented:
Hey Robyt3, good PR overall, I just have a few issues with it that'd I like to point at.

* Including screenshots when doing visual stuff is much appreciated and very easy to do on github. That way we can know what changed instantly and if it looks wrong or not.
* I feel like moving code over to another file must have some kind of purpose. menus.cpp is the "menu" module and includes all the ui functions you need to build menus, I'm not sure creating small files in the hope that we'll reuse them later is a good idea. However extracting the menu functions to be usable for the main menu as well as the editor would be great.
* Finally the last point is more subtle. Moving functions to be methods breaks the pattern we use to make menus. Usually we define a handle/object like `CButtonContainer` as static and pass it to a function. We do it that way because as you can see here: https://github.com/teeworlds/teeworlds/pull/2396/files#diff-a3a89c00bd80fbe33f82d7bc1b8aaf0aR107, a pointer is used as UI item ID. Your changes break this pattern and may result in someone declaring CListBox as non static thus making it non functional.

Anyway this is just my opinion of course, in the end it is up to @oy to decide.

Robyt3 commented:
I appreciate the feedback,
- Yeah, I generally agree, but there wasn't anything visually that actually changed. The new messages are displayed just like the others and "restoring the selected state" isn't something I can capture in screenshots very well.
- Moving code to different files (like the demo and browser menus code already are) improves compilation time because only changed files need to be recompiled. It also makes it easier to find the associated functions. Right now you still include all the UI blocks with menus.h, but menus.cpp is rather convoluted and contains not only UI components but also the main menu framework and some settings UI.
- The problem would have occurred if someone declares a CListBox struct as non static. Something like this should be part of the documentation, and probably also part of the source code, so users of a class know how to use it.

oy commented:
First about this PR:
There's some strange behaviour. When I start the client "ui_server_address" is cleared although the address that was in there on start is in the list.
When you enable more filter settings, so your selection is hidden, it selects some new random? entry.

About the UI:
I thought about stripping UI elements before. The menu files build up the menu pages and UI elements like buttons, list boxes etc. could be extracted from there and be put in a UI element file structure. Then it would be easier to reuse them like mentioned above.
The UI item ID needs to be unique for an item and shall not change. So we used static addresses for that. We could change that to a function which provides a unique ID f.e. Would be a different Issue though :)

Robyt3 commented:
What's the appropriate behaviour when you change the filter? Right now it keeps the selection index but changes the address to reflect the item that has that index now. Not really optimal, but what else is supposed to happen? Server address should reflect the selection, if the previous selection is no longer possible should it completely unselect?

oy commented:
The address in "server address" should stay there unless you actually pick another entry in the list. When it isn't in the list, it would probably be best to not select an entry.

Robyt3 commented:
Should all be fixed now.

Also added a "Fetching server info" message to show while no servers are available during refresh.

oy commented:
When you type something in the search field it still selects a different entry.
And when you activate another filter your server address entry gets erased. That's kinda odd. And you can have a different selection for each filter, though it will save the selected one from the last filter you used in `ui_server_address` on exit.

Robyt3 commented:
It shouldn't erase the address anymore and also work with sorting the list.

Can you describe the issue you have with `ui_server_address`? Should we store multiple addresses for the different filters?

oy commented:
Hm, not sure. Either keep the entry for all filters or have a selection for every filter and then probably save it with the filter in ui_settings.json. Dunno what the best option is. Opinions?

Robyt3 commented:
Now there will only be one selected server between all filters, determined by the address.

I changed the ui_settings.json to store the expanded filter for each browser page separately.

oy commented:
This changes the json format, so when you run it the first time all filters are collapsed. Could you expand the "All" filter then, to keep the current behavior.
Also when you switch from global to lan tab and no filter is expanded it deletes ui_server_address.
Dunno if it should clear the server address when switching to lan.

Robyt3 commented:
> This changes the json format, so when you run it the first time all filters are collapsed. Could you expand the "All" filter then, to keep the current behavior.

I'll adjust that.

> Also when you switch from global to lan tab and no filter is expanded it deletes ui_server_address.
Dunno if it should clear the server address when switching to lan.

I think that should be part of #2003, so switching tabs doesn't keep an address for servers which are not shown in the new tab, but I can change it as well, if you want to.

oy commented:
But if you change back to global the server adress isn't coming back.

Maybe add a ui_lan_server_address config? Then you don't run into problems when shutting down the client when lan tab is open and the server address is empty.

Robyt3 commented:
> Maybe add a ui_lan_server_address config? Then you don't run into problems when shutting down the client when lan tab is open and the server address is empty.

That should work. I'll implement that tomorrow.

Robyt3 commented:
Now there are `ui_internet_server_address` and `ui_lan_server_address`. Default for both is "localhost:8303", not sure if there is a better alternative.

I removed how connecting via console command changes the UI address, because the engine client doesn't know about the menus. The ingame menu now displays the actual address of the server info and not the UI address.

I also rebased with the favorites password changes which broke with this.

Dune-jr commented:
Tested, looks good.
Very nitpicky, but I would suggest `ui_server_address` and `ui_server_address_lan` in case anyone used the setting in their config.
